src/winter_onboarding_2021/fleet_management_service/handlers/fleet.clj

```diff
@@ -1,31 +1,61 @@
             [winter-onboarding-2021.fleet-management-service.config :as config]
             [winter-onboarding-2021.fleet-management-service.views.fleet :as views]
             [winter-onboarding-2021.fleet-management-service.models.fleet :as fleet-model]
+            [winter-onboarding-2021.fleet-management-service.models.user :as user-models]
             [winter-onboarding-2021.fleet-management-service.models.users-fleets :as users-fleets-models]))
 
 
-(defn create-fleet [{:keys [user form-params]}]
-  (let [user-id (:users/id user)
-        fleet-name (:name form-params)
-        fleet-data (utils/namespace-keys :fleets {:name fleet-name
-                                                  :created-by user-id})]
-    (if (s/valid? ::specs/fleets fleet-data)
-      (let [fleet (fleet-model/create fleet-data)
-            fleet-id (:fleets/id fleet)
-            _ (users-fleets-models/create user fleet)]
-        (->  (utils/flash-msg "Fleet created successfully!" true)
-             (merge (response/redirect (format "/fleets/%s" (str fleet-id))))))
-      (->  (utils/flash-msg "Could not create fleet, try again!" false)
-           (merge (response/redirect "/fleets/new"))))))
+(defn- wrap-in-vector [entity]
+  (if (not (vector? entity))
+    (vector entity)
+    entity))
 
-(defn new [_]
-  {:title "Create fleet"
-   :content (views/create-fleet)})
+(defn create-fleet [{admin :user form-params :form-params}]
+  (jdbc/with-transaction [tx db/db-conn]
+    (let [{admin-id :users/id
+           org-id :users/org-id} admin
+          {fleet-name :name
+           :keys [managers drivers]} form-params
+          fleet-data #:fleets{:name fleet-name
+                              :created-by admin-id
+                              :org-id org-id}
+          managers-uuids (map utils/string->uuid (wrap-in-vector managers))
+          drivers-uuids (map utils/string->uuid (wrap-in-vector drivers))
+          team-members (remove nil? `(~@managers-uuids ~@drivers-uuids))
+
+          ;; If managers or drivers are empty, it's okay to create a fleet
+          managers-in-org? (or (empty? managers)
+                               (user-models/users-in-org? tx managers-uuids {:organisations/id org-id}))
+          drivers-in-org?  (or (empty? drivers)
+                               (user-models/users-in-org? tx drivers-uuids {:organisations/id org-id}))]
+      
+      (if (and (s/valid? ::specs/fleets fleet-data)
+               managers-in-org?
+               drivers-in-org?)
+        (let [{fleet-id :fleets/id :as fleet} (fleet-model/create tx fleet-data)]
+          (users-fleets-models/create tx admin fleet)
+          (doall (map #(users-fleets-models/create tx {:users/id %} fleet)
+                      team-members))
+          (->  (utils/flash-msg "Fleet created successfully!" true)
+               (merge (response/redirect (format "/fleets/%s" (str fleet-id))))))
+        (->  (utils/flash-msg "Could not create fleet, try again!" false)
```

Comment:
There are multiple reasons, a fleet could not be created. Example:- If the connection is lost, if any one of the members(driver or manager) does not belong to the same org, or data is not valid the flash message would say ` Could not create fleet`. We did not want to expose the reason for which it fails, which will expose the implementation detail to the end user, since this is at handler layer. So we added a generic error message.

Reply:
Some errors could be  informative to the user. Some part of implementation detail, which is obvious, could be shared.